File: solutions/python/luhn/2/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num
        self.pos = 0

    def valid(self):
        self.pos = 0
        if any(not ch.isdigit() and ch!=' ' for ch in self.card_num):
            return False
        mysum = sum([self.luhn_digit(ch) for ch in self.card_num[::-1]])
        return self.pos>1 and mysum%10==0

    def luhn_digit(self,ch):
        if not ch.isdigit():
            return 0
        self.pos += 1
        x = int(ch)
        if self.pos%2==0 and x!=9:
            x = 2*x % 9
        return x
       
            

File: solutions/cpp/luhn/2/luhn.cpp
#include "luhn.h"
#include <string>
#include <algorithm>

using std::string;

namespace luhn {
    bool valid (string number){
       int sum=0;
       char c;     
       auto end_pos = remove(begin(number), end(number), ' ');
       number.erase(end_pos, end(number));
       if (number.length()<=1)
            return false;
        for (int i=number.length()-1; i>=0; i--){
                c = number.at(i);
                if (isdigit(c)){
                   int d = (int(c) - 48);
                   if (!((number.length() - i)%2)) 
                       (d!=9) ? (d = 2*d % 9) : (d=9);
                   sum += d;
                } else 
                  return false;                 
        }
        return !(sum % 10);
    }
}  // namespace luhn
